File: src/main/java/com/yourway/alert/streaming/JobAlert.java
package com.yourway.alert.streaming;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.yourway.alert.domain.Query;
import com.yourway.alert.streaming.settings.Settings;
import com.yourway.alert.utils.UtilKafka;
import org.apache.commons.lang3.math.NumberUtils;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.api.java.UDF6;
import org.apache.spark.sql.streaming.StreamingQuery;
import org.apache.spark.sql.streaming.StreamingQueryException;
import org.apache.spark.sql.types.DataTypes;
import scala.collection.JavaConverters;
import scala.collection.Seq;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Properties;
import java.util.concurrent.TimeoutException;

import static org.apache.spark.sql.functions.*;
import static org.apache.spark.sql.types.DataTypes.StringType;

public class JobAlert {

    public static void main(String[] args) throws TimeoutException, StreamingQueryException {
        SparkSession spark = SparkSession
                .builder()
                .appName("Alert Job Yourway")
                .getOrCreate();
        spark.sparkContext().setLogLevel("ERROR");
        Dataset<Row> df = spark
                .readStream()
                .format("kafka")
                .option("kafka.bootstrap.servers", Settings.KAFKA_URI)
                .option("subscribe", Settings.TOPIC_JOB)
                .load();
        Seq<String> seq = JavaConverters.asScalaIteratorConverter(Settings.FIELD_JOB.iterator()).asScala().toSeq();

        Dataset<Row> data = df.select(
                json_tuple(
                        col("value").cast(StringType), seq
                )
        );
        data.printSchema();
        // get user query from kafka
        Properties props = UtilKafka.createConsumer("user");
        ConsumerRecord<String, String> mess = UtilKafka.getLatestMessage(props, Settings.TOPIC_USER_QUERY);
        Gson gson = new Gson();
        Type type = new TypeToken<ArrayList<Query>>() {
        }.getType();
        if (mess != null) {
            ArrayList<Query> listQuery = gson.fromJson(mess.value(), type);
            // matching
            try {
                UDF6<String, String, String, String, String, String, String> check_matching = (address, age, salary, yearExperiments, eduLevel, jobAttribute) -> {
                    List<String> user_id = new ArrayList<>();
                    int min_age = 0;
                    int max_age = 200;

                    if (age != null && !age.isEmpty()) {
                        String[] array_age = age.split(",");
                        if (array_age.length > 1) {
                            min_age = Settings.DICT_MIN_AGE.get(array_age[0]);
                            max_age = Settings.DICT_MAX_AGE.get(array_age[array_age.length - 1]);
                        }
                    }
                    for (Query query : listQuery) {
                        Long querySalary = query.getSalary();

                        boolean is_number_salary = true;
                        if (salary != null && !salary.isEmpty()) {
                            float salaryFloat = Float.parseFloat(salary);
                            is_number_salary = salaryFloat >= querySalary;
                        }
                        boolean is_not_none_address = false;
                        if (address != null && !address.isEmpty()) {
                            is_not_none_address = address.contains(query.getCompanyAddress());
                        }

                        try {
                            if (is_not_none_address && min_age <= query.getAge() && query.getAge() <= max_age &&
                                    is_number_salary && Objects.equals(query.getYearExperiences(), yearExperiments) &&
                                    Objects.equals(query.getEducationLevel(), eduLevel) &&
                                    Objects.equals(query.getJobAttribute(), jobAttribute)) {
                                user_id.add(String.valueOf(query.getUserId()));
                            }
                        } catch (Exception e) {
                            System.out.println(e.getMessage());
                        }
                    }
                    return String.join(",", user_id);
                };
                spark.sqlContext().udf().register("matching", check_matching, DataTypes.StringType);
                data = data.withColumn("value",
                        callUDF("matching",
                                col("c2"), col("c14"), col("c10"), col("c22"), col("c15"), col("c19")));
            } catch (Exception e) {
                System.out.println(e.getMessage());
            }
        } else {
            data = data.withColumn("value", lit(null).cast(StringType));
        }

        data = data.withColumn(
                "key", col("c0")
        );
        data = data.filter(col("value").isNotNull());

        StreamingQuery ds = data
                .writeStream()
                .format("kafka")
                .option("kafka.bootstrap.servers", Settings.KAFKA_URI)
                .option("kafka.group.id", Settings.GROUP_ID_USER)
                .option("checkpointLocation", Settings.CHECKPOINT_PATH)
                .option("topic", Settings.TOPIC_USER)
                .start();

        ds.awaitTermination();
    }
}

File: streaming/JobAlert.py
import csv
import json
import random

from pyspark.sql import SparkSession
from pyspark.sql.functions import udf, expr, json_tuple, from_json, col
from pyspark.sql.types import StringType, StructType, IntegerType, StructField, DateType, LongType, FloatType, Row

from settings import KAFKA_URI, DICT_MAX_AGE, DICT_MIN_AGE, \
    TOPIC_JOB, TOPIC_USER_QUERY, FIELD_JOB, FIELD_QUERY

decode_string = udf(lambda x: x.decode('utf-8'), StringType())


def matching(queries, address: str, age=None, salary=None, year=None, edu_level=None, job_attribute=None):
    user_id = []
    if age is not None:
        array_age = age.split(',')
        if len(array_age) > 1:
            min_age = int(DICT_MIN_AGE[array_age[0]])
            max_age = int(DICT_MAX_AGE[array_age[-1]])
        else:
            min_age = 0
            max_age = 200
    else:
        min_age = 0
        max_age = 200

    # for query in queries:
    #     # is_number_salary = True
    #     # if salary is not None:
    #     #     is_number_salary = query['salary'] >= float(salary)
    #     #
    #     # is_not_none_address = False
    #     # if address is not None:
    #     #     is_not_none_address = query['company_address'] in address
    #     # try:
    #     #     # if is_not_none_address and min_age <= int(query[2]) <= max_age and is_number_salary and \
    #     #     #         query[4] == year and query[5] == edu_level and query[6] == job_attribute:
    #     #     if min_age <= int(query['age']) <= max_age and is_number_salary:
    #     #         user_id.append(
    #     #             dict(user_id=query['user_id'], contact=query['contact'])
    #     #         )
    #     # except Exception as e:
    #     #     logging.error(e)
    #     for i in range(1000):
    #         a = 1 + 1
    #     user_id.append(
    #         dict(user_id=query['user_id'], contact=query['contact'])
    #     )

    # return '[' + ','.join(str(x) for x in user_id) + ']'
    return str(random.randint(1, 100000))


def read_data():
    data = []
    with open('user_query.csv', encoding='utf-8') as f:
        csv_reader = csv.DictReader(f)
        for row in csv_reader:
            data.append(row)
    return json.loads(json.dumps(data, indent=4))


def main():
    spark = SparkSession \
        .builder \
        .appName("Job Alert Yourway") \
        .getOrCreate()
    spark.sparkContext.setLogLevel("ERROR")
    spark.conf.set("spark.sql.execution.arrow.pyspark.enabled", "true")
    spark.conf.set("spark.sql.crossJoin.enabled", "true")
    df_job = spark \
        .readStream \
        .format("kafka") \
        .option("kafka.bootstrap.servers", KAFKA_URI) \
        .option("subscribe", TOPIC_JOB) \
        .load()
    df_query = spark \
        .readStream \
        .format("kafka") \
        .option("kafka.bootstrap.servers", KAFKA_URI) \
        .option("subscribe", TOPIC_USER_QUERY) \
        .load()
    schema_job = StructType([
        StructField("id", LongType()),
        StructField("application_deadline", DateType()),
        StructField("company_address", IntegerType()),
        StructField("salary", FloatType()),
        StructField("ages", StringType()),
        StructField("education_level", StringType()),
        StructField("position", StringType()),
        StructField("job_attribute", StringType()),
        StructField("year_experiences", FloatType()),
    ])
    schema_query = StructType([
        StructField("query_id", LongType()),
        StructField("query_user_id", LongType()),
        StructField("query_company_address", StringType()),
        StructField("query_job_role", StringType()),
        StructField("query_age", IntegerType()),
        StructField("query_salary", FloatType()),
        StructField("query_year_experiences", FloatType()),
        StructField("query_education_level", StringType()),
        StructField("query_job_attribute", StringType()),
        StructField("query_contact", StringType()),
        StructField("query_group_by", StringType()),
        StructField("query_slide_window", StringType()),
    ])

    data_job = df_job.withColumn(
        "data", from_json(col("value").astype(StringType()), schema_job)
    ).select("key", "offset", "partition", "timestamp", "timestampType", "topic", "data.*")

    data_query = df_query.withColumn(
        "data", from_json(col("value").astype(StringType()), schema_query)
    ).select("key", "offset", "partition", "timestamp", "timestampType", "topic", "data.*")

    data = data_job.alias("JOB").join(data_query.alias("QUERY"), expr("""
        QUERY.query_job_attribute = JOB.job_attribute AND
        QUERY.query_job_role = JOB.position AND
        JOB.education_level = QUERY.query_education_level AND
        JOB.company_address like concat("%", QUERY.query_company_address, "%") AND
        JOB.salary >= QUERY.query_salary AND
        QUERY.timestamp <= JOB.timestamp + interval 1 hour AND
        QUERY.timestamp > JOB.timestamp
    """), "full_outer")

    # check_matching = udf(
    #     lambda address, age, salary, year, edu_level, job_attribute:
    #     matching(address, age, salary, year, edu_level, job_attribute), StringType()
    # )
    #
    # # match
    # data = data.withColumn(
    #     "value", check_matching(data["company_address"], data["ages"], data["salary"], data["year_experiences"],
    #                             data["education_level"], data["job_attribute"])
    # )
    #
    # data = data.withColumn(
    #     "key", col("id")
    # )
    #
    # data = data.filter(col("value").isNotNull())

    # run app
    # app = data \
    #     .writeStream \
    #     .format("kafka") \
    #     .option("kafka.bootstrap.servers", KAFKA_URI) \
    #     .option("checkpointLocation", CHECKPOINT_PATH) \
    #     .option("topic", TOPIC_USER) \
    #     .start()
    app = data \
        .writeStream \
        .format('console') \
        .start()

    app.awaitTermination()


if __name__ == '__main__':
    main()
